✏️Add quickstart (#435)
kids-first/kf-lib-data-ingest: docs/source/quickstart.rst

.. _Quickstart:

**********
Quickstart
**********

Generate a new ingest package
=============================

> The first thing to do after installing the ingest library CLI

Reviewer: ```suggestion
Each study that needs to be ingested will have a single directory that contains
all of the configuration needed by the ingest pipeline to extract, transform,
and load the study into a target data store or service. This is known as the ingest package.

The first thing to do after installing the ingest library CLI
```

> is to create a new package named ``my_study``::

Reviewer: ```suggestion
is to create a new ingest package named ``my_study``::
```

$ kidsfirst new --dest_dir=my_study

We can run it and test to make sure everything works by using the following
command

.. code-block:: text

    $ kidsfirst test my_study

The logs on your screen should indicate a successful test run like this:

.. code-block:: text

    ======================================== 1 passed in 0.12s ============================================================
    2020-02-18 17:40:36,039 - DataIngestPipeline - Thread: MainThread - INFO - ✅ User defined data validation tests passed
    2020-02-18 17:40:36,039 - DataIngestPipeline - Thread: MainThread - INFO - END data ingestion
    2020-02-18 17:40:36,039 - kf_lib_data_ingest.app.cli - Thread: MainThread - INFO - ✅ Ingest pipline passed validation!

> This package has a simple test piece of data to extract from. You need to write an additional extraction

Reviewer: Remove these two lines, we'll introduce the extract configs in the next subsection. See below suggestion

script for each file that you'd like to ingest. Let's look at an example of doing that.

Modify a template extract config
================================

> 

Reviewer: ```suggestion
An extract config file does 3 things to the source data:

1. Select or extract the desired subset of data from the source data
2. Clean the selected data (remove trailing whitespaces, etc)
3. Map the cleaned data's attributes and values to Kids First entity attributes
   and acceptable values.
```

> We're going to extract data from the following source data file (different

Reviewer: ```suggestion
We're going to create an additional extract config to extract data from the following source data file (different
```

                },
                value_name=CONCEPT.PHENOTYPE.OBSERVED,
                map_for_values=observed_yes_no
            )
        ]
    ]

It goes in the ``extract_configs`` directory like this::

    my_study/
    ├── data
    │   └── clinical.tsv
    └── extract_configs
        ├── extract_config.py
        └── family_and_phenotype.py

Run the ingest pipeline
=======================

The Warehouse Server
--------------------

Data that we ingest is meant to be shared, not stored locally on your machine!
It's easy to configure the ingest system to connect to a remote warehousing database
which will centrally store the relevant data.

Local Environment Configuration
-------------------------------

All that's needed is the appropriate database to be setup with the required access
permissions (see :ref:`Warehousing` for all the details). We just need to set the local variable
``KF_WAREHOUSE_DB_URL`` with a complete login URL::

    export KF_WAREHOUSE_DB_URL=postgresql://<username>:<password>@<address>:<port>/postgres

The final Extraction product
----------------------------

Now we're ready to perform extraction, which is done with the following command::

>     $ kidsfirst ingest <your package>

Reviewer: ```suggestion
    $ kidsfirst ingest --no_warehouse my_study
```

We're just seeing how this package works on dummy data, so we don't want

> to actually upload anything to a warehouse. This won't happen if we use the ``test``

Reviewer: ```suggestion
to actually upload anything to a warehouse. This won't happen if we use the ``--no_warehouse`` option
```

command, like so::

    $ kidsfirst test my_study

> After completing this extraction that we specified using the script,

Reviewer: I don't think we need to show the extracted table here. You already mention where ingest stage output gets written. I think thats enough

    "0", "1", "2", "3", "Female", "SP001B", "0.166666667", "Extra eardrum", "Negative"
    "1", "2", "", "", "", "SP002A", "18.125", "Extra eardrum", "Negative"
    "1", "2", "", "", "", "SP002B", "18.125", "Extra eardrum", "Negative"
    "2", "3", "", "", "", "SP003A", "1.416666667", "Extra eardrum", "Negative"
    "2", "3", "", "", "", "SP003B", "1.416666667", "Extra eardrum", "Negative"
    "3", "4", "5", "6", "Male", "SP004A", "0.166666667", "Extra eardrum", "Negative"
    "3", "4", "5", "6", "Male", "SP004B", "0.166666667", "Extra eardrum", "Negative"
    "4", "5", "", "", "", "SP005A", "1.416666667", "Extra eardrum", "Negative"
    "4", "5", "", "", "", "SP005B", "1.416666667", "Extra eardrum", "Negative"
    "5", "6", "", "", "", "SP006", "1814.375", "Extra eardrum", "Negative"
    "6", "7", "8", "9", "Male", "SP007", "0.208333333", "Extra eardrum", "Positive"
    "7", "8", "", "", "", "SP008A", "2.166666667", "Extra eardrum", "Positive"
    "7", "8", "", "", "", "SP008B", "2.166666667", "Extra eardrum", "Positive"
    "8", "9", "", "", "", "SP009A", "1.041666667", "Extra eardrum", "Positive"
    "8", "9", "", "", "", "SP009B", "1.041666667", "Extra eardrum", "Positive"

> 

Reviewer: Can you add a section for the transform stage as well? Provide a 1-2 line explanation of what the transform module does (joins extracted output tables together) and then display the transform module and instruct the user to uncomment the merge code.

Generate a new ingest package
=============================

Each study that needs to be ingested will have a single directory that contains
all of the configuration needed by the ingest pipeline to extract, transform,
and load the study into a target data store or service. This is known as the
ingest package.

The first thing to do after installing the ingest library CLI
is to create a new ingest package named ``my_study``::

$ kidsfirst new --dest_dir=my_study

We can run it and test to make sure everything works by using the following
command

.. code-block:: text

    $ kidsfirst test my_study

The logs on your screen should indicate a successful test run like this:

.. code-block:: text

    ======================================== 1 passed in 0.12s ============================================================
    2020-02-18 17:40:36,039 - DataIngestPipeline - Thread: MainThread - INFO - ✅ User defined data validation tests passed
    2020-02-18 17:40:36,039 - DataIngestPipeline - Thread: MainThread - INFO - END data ingestion
    2020-02-18 17:40:36,039 - kf_lib_data_ingest.app.cli - Thread: MainThread - INFO - ✅ Ingest pipline passed validation!

> 

Reviewer: ```suggestion

So far your ingest package is pretty basic. It has 1 source data file and the Python modules needed to extract, clean and transform the data for the target service. If you inspect it, it will look like this::

    my_study/
    ├── ingest_package_config.py
    ├── transform_module.py
    ├── data
    │   └── clinical.tsv
    ├── extract_configs/
    │   └── extract_config.py
    └── tests
        ├── conftest.py
        └── test_custom_counts.py
```

                in_col=6,  # age (hrs) (first)
                m=lambda x: int(x) / 24,
                out_col=CONCEPT.PHENOTYPE.EVENT_AGE_DAYS
            ),
            melt_map(
                var_name=CONCEPT.PHENOTYPE.NAME,
                map_for_vars={
                    "CLEFT_EGO": "Cleft ego",
                    "CLEFT_ID": "Cleft id"
                },
                value_name=CONCEPT.PHENOTYPE.OBSERVED,
                map_for_values=observed_yes_no
            )
        ],
        [
            value_map(
                in_col=9,  # age (hrs) (second)
                m=lambda x: int(x) / 24,
                out_col=CONCEPT.PHENOTYPE.EVENT_AGE_DAYS
            ),
            melt_map(
                var_name=CONCEPT.PHENOTYPE.NAME,
                map_for_vars={
                    "EXTRA_EARDRUM": "Extra eardrum"
                },
                value_name=CONCEPT.PHENOTYPE.OBSERVED,
                map_for_values=observed_yes_no
            )
        ]
    ]

It goes in the ``extract_configs`` directory like this::

    my_study/
    ├── data
    │   └── clinical.tsv
    └── extract_configs
        ├── extract_config.py
        └── family_and_phenotype.py

> Run the ingest pipeline

Reviewer: Can you move this whole section (Run the ingest pipeline) to the end after the transform section

=======================

The Warehouse Server
--------------------

Data that we ingest is meant to be shared, not stored locally on your machine!
It's easy to configure the ingest system to connect to a remote warehousing
database which will centrally store the relevant data.

Local Environment Configuration
-------------------------------

All that's needed is the appropriate database to be setup with the required
access permissions (see :ref:`Warehousing` for all the details). We just need
to set the local variable
``KF_WAREHOUSE_DB_URL`` with a complete login URL::

    export KF_WAREHOUSE_DB_URL=postgresql://<username>:<password>@<address>:<port>/postgres

> The final Extraction product

Reviewer: I don't think we need this little section. Can you please remove it?

----------------------------

> Now we're ready to perform extraction, which is done with the following

Reviewer: Let's move this (introducing the ingest command) to the end of this Run ingest pipeline section. And instead of saying `Now we're ready to perform extraction, which is done with the following` can you say something like:

```
If the ingest package passes validation with the test command and you're happy with the output, you're ready ingest the data into your target service using the ingest command::

    $ kidsfirst ingest my_study 

The library defaults to ingesting data into the Kids First Data Service at the base url: `http://localhost:5000`. You can change the base URL for this service using the `--target_url` CLI option.
```

command::

    $ kidsfirst ingest <your package>

> We're just seeing how this package works on dummy data, so we don't want

Reviewer: ```suggestion
We're ready to test the ingest package again and since we're just seeing how this package works on dummy data, we don't want
```

to actually upload anything to a warehouse. This won't happen if we use the
``--no_warehouse`` option command, like so::

>     $ kidsfirst test my_study

Reviewer: ```suggestion
    $ kidsfirst test --no_warehouse my_study
```

Debugging
=========

Run single stage
----------------

The entire ingest pipeline consists of the familiar ETL steps (extract,
transform, and load). The package allows any logical subset of these three
stages to be executed instead (output from the previous stages must exist in
order for extract or transform to be run). This can be done using the
``--stages`` argument with the char sequence ``etl``.

View ingest log
---------------

The pipeline logs messages to the console and also stores them in a log file,
which is by default located at ``<ingest package dir>/logs/ingest.log``.

View single stage output
------------------------

Every ingest stage has the option to write its output to a directory that
follows this path pattern: ``<ingest package dir>/output/<name of stage>``.

> Transform Stage

Reviewer: ```suggestion
Modify the transform module
```

Each study that needs to be ingested will have a single directory that contains
all of the configuration needed by the ingest pipeline to extract, transform,
and load the study into a target data store or service. This is known as the
ingest package.

The first thing to do after installing the ingest library CLI
is to create a new ingest package named ``my_study``::

$ kidsfirst new --dest_dir=my_study

We can run it and test to make sure everything works by using the following
command

.. code-block:: text

    $ kidsfirst test my_study

The logs on your screen should indicate a successful test run like this:

.. code-block:: text

    ======================================== 1 passed in 0.12s ============================================================
    2020-02-18 17:40:36,039 - DataIngestPipeline - Thread: MainThread - INFO - ✅ User defined data validation tests passed
    2020-02-18 17:40:36,039 - DataIngestPipeline - Thread: MainThread - INFO - END data ingestion
    2020-02-18 17:40:36,039 - kf_lib_data_ingest.app.cli - Thread: MainThread - INFO - ✅ Ingest pipline passed validation!

So far your ingest package is pretty basic. It has 1 source data file and

> the Python modules needed to extract, clean and transform the data for the

Reviewer 2: ```suggestion
the Python modules needed to extract, clean, and transform the data for the
```

Author: i can't believe i missed on oxford comma...

target service. If you inspect it, it will look like this::

    my_study/
    ├── ingest_package_config.py
    ├── transform_module.py
    ├── data
    │   └── clinical.tsv
    ├── extract_configs/
    │   └── extract_config.py

>     └── tests

Reviewer 2: inconsistent use of trailing slash on directory names

            out_col=CONCEPT.BIOSPECIMEN.ID
        ),
        [
            value_map(
                in_col=6,  # age (hrs) (first)
                m=lambda x: int(x) / 24,
                out_col=CONCEPT.PHENOTYPE.EVENT_AGE_DAYS
            ),
            melt_map(
                var_name=CONCEPT.PHENOTYPE.NAME,
                map_for_vars={
                    "CLEFT_EGO": "Cleft ego",
                    "CLEFT_ID": "Cleft id"
                },
                value_name=CONCEPT.PHENOTYPE.OBSERVED,
                map_for_values=observed_yes_no
            )
        ],
        [
            value_map(
                in_col=9,  # age (hrs) (second)
                m=lambda x: int(x) / 24,
                out_col=CONCEPT.PHENOTYPE.EVENT_AGE_DAYS
            ),
            melt_map(
                var_name=CONCEPT.PHENOTYPE.NAME,
                map_for_vars={
                    "EXTRA_EARDRUM": "Extra eardrum"
                },
                value_name=CONCEPT.PHENOTYPE.OBSERVED,
                map_for_values=observed_yes_no
            )
        ]
    ]

It goes in the ``extract_configs`` directory like this::

    my_study/
    ├── data
    │   └── clinical.tsv

>     └── extract_configs

Reviewer 2: inconsistent use of trailing slash on directory names

        ├── extract_config.py
        └── family_and_phenotype.py

Debugging
=========

Run single stage
----------------

The entire ingest pipeline consists of the familiar ETL steps (extract,
transform, and load). The package allows any logical subset of these three
stages to be executed instead (output from the previous stages must exist in
order for extract or transform to be run). This can be done using the

> ``--stages`` argument with the char sequence ``etl``.

Reviewer 2: ```suggestion
``--stages`` argument with a substring of the char sequence ``etl``.
```

        output dict. They key must match the name of an existing target concept.
        The value will be the DataFrame to use when building instances of the
        target concept.

        A typical example would be:

        {
            'family_relationship': family_relationship_df,
            'default': all_merged_data_df
        }

        """
        df = mapped_df_dict["extract_config.py"]

        # df = outer_merge(
        #     mapped_df_dict['extract_config.py'],
        #     mapped_df_dict['family_and_phenotype.py'],
        #     on=CONCEPT.BIOSPECIMEN.ID,
        #     with_merge_detail_dfs=False
        # )

        return {DEFAULT_KEY: df}

The interesting part that we need is the block at the bottom that's commented
out. Uncomment this block and then run ``kidsfirst test`` to perform
extraction and tranformation.

Run the ingest pipeline
=======================

The Warehouse Server
--------------------

Data that we ingest is meant to be shared, not stored locally on your machine!
It's easy to configure the ingest system to connect to a remote warehousing
database which will centrally store the relevant data.

> The final Extraction product

Reviewer: ```suggestion
```

🔥Rm this line please

----------------------------

> We're ready to test the ingest package again and since we're just seeing how

Reviewer 2: ```suggestion
We're ready to test the ingest package again, and, since we're just seeing how
```

this package works on dummy data, we don't want
to actually upload anything to a warehouse. This won't happen if we use the
``--no_warehouse`` option command, like so::

    $ kidsfirst test --no_warehouse my_study

If the ingest package passes validation with the test command and you're happy

> with the output, you're ready ingest the data into your target service using

Reviewer 2: ```suggestion
with the output, you're ready to ingest the data into your target service using
```
